### PR TITLE
improve keyboard accessibility inside modals

### PR DESCRIPTION
**Pull request by ETLaurent**
⏬ Don't forget to read the "**Other information**" section at the end ⏬

❓ What is expected from the review ?

- implementation system looks 🆗 with everyone (DX-friendly, no performance issue...)
- tips and things that might have been forgotten in the PR

## Summary

- Trap the focus in modals by default (except for the `AposCommandMenuShortcut.vue` which explicitly disables it with `trapFocus: false`)
- Create `AposFocusMixin.js`, inspired by the existing `trapFocus()` code from `AposModal.vue`, that has been improved and now provides methods to better handle the focus with the keyboard.

In this PR, this **mixin is used** in `AposModal.vue`, to retain ("trap") the focus inside the current modal, and give it back to the previous modal when closing it.

`AposModal.vue` exposes the `modal.triggerFocusRefresh` prop which refreshes DOM elements to focus when incremented. It is useful to get the new or remove the old elements after an update made inside a modal, which adds/remove DOM elements, like an XHR call to get the pieces list in the AposDocsManager modal, for instance:

```js
await this.getPages();
this.modal.triggerFocusRefresh++;
```

`AposButtonSplit.vue` also **uses the mixin**, as a way of testing the robustness of the methods it provides, in another context (not for modals).
It allows the focus to be retained on the menu items, in order to cycle through them, and give the focus back to the previous focused element of the modal when an item is selected:

image

Finally, in this PR we changed the manual event listening to use native Vue listeners, which avoid using `refs` too much and let Vue handle the event listeners removal when unmounted:

```js
// in AposModal and AposButtonSplit
@keydown="cycleElementsToFocus"

// in AposModal
@focus.capture="storeFocusedElement"
```

**Other changes:**

- Items in expanded menu to add widgets are now focusable
- The drag&drop area in `AposMediaUploader.vue` is now focusable instead of the hidden file `input`, and pressing `enter` or `space` on it will trigger a `click` on the actual file `input`:
image
- Improve and fix DOM selectors to get visible focusable elements, and respect the `tabindex` W3C rule.
- Improve focus on the `AposModalShareDraft.vue` modal (clicking on focused elements will be entirely working in another PR)
- Add temporary commented styles (`outline` and `background-color`) for debugging

---

useful sources:

https://www.w3.org/WAI/test-evaluate/preliminary/#interaction
https://uxdesign.cc/how-to-trap-focus-inside-modal-to-make-it-ada-compliant-6a50f9a70700
https://css-tricks.com/a-css-approach-to-trap-focus-inside-of-an-element/
https://github.com/medialize/ally.js/blob/master/src/maintain/tab-focus.js
https://github.com/gdkraus/accessible-modal-dialog/blob/master/modal-window.js
https://github.com/ghosh/Micromodal/blob/master/lib/src/index.js#L3

## What are the specific steps to test this change?

Use testbed or any other a3 project.
Link apostrophe (on the `pro-3625-accessibility` branch).

Navigate inside pages/docs manager modals and pieces modals using the keyboard.
Tabs should be focusable.
Closing a modal should give the focus back to the last focused element of the parent modal.

## What kind of change does this PR introduce?
*(Check at least one)*

- [ ] Bug fix
- [ ] New feature
- [ ] Refactor
- [ ] Documentation
- [ ] Build-related changes
- [x] Other

## Make sure the PR fulfills these requirements:

- [x] It includes a) the existing issue ID being resolved, b) a convincing reason for adding this feature, or c) a clear description of the bug it resolves
- [x] The changelog is updated
- [ ] Related documentation has been updated
- [ ] Related tests have been updated

If adding a new feature without an already open issue, it's best to open a **feature request issue** first and wait for approval before working on it.

**Other information:**

This needs to be thoroughly tested, since _a lot_ of use-cases exist in terms of keyboard navigation, and this PR mainly establishes a **base focus system** that will surely need to be improved when continuing the work on keyboard accessibility.

In a lot of cases, the focus is still lost, or ~~trapped~~ stuck on an element, which in terms of accessibility, is **bad**.
Tickets have been created for issues that have been spotted, but there will probably be more to come as we will be testing Apostrophe with the keyboard only.

This is a long term endeavour.

### CODE REVIEW

File: modules/@apostrophecms/modal/ui/apos/components/AposModal.vue [hunk 1]
```diff
@@ -232,48 +246,76 @@ export default {
...
       // Adapted from https://uxdesign.cc/how-to-trap-focus-inside-modal-to-make-it-ada-compliant-6a50f9a70700
       // All the elements inside modal which you want to make focusable.
-      const focusableElements = [
+      const elementSelectors = [
         'button',
         '[href]',
         'input',
         'select',
         'textarea',
-        '[tabindex]:not([tabindex="-1"]'
```
Comment (ETLaurent) [inline modules/@apostrophecms/modal/ui/apos/components/AposModal.vue:243]: missing `)`, _that_ was hard to find why it was not working...

File: modules/@apostrophecms/area/ui/apos/components/AposAreaExpandedMenu.vue [hunk 2]
```diff
@@ -54,7 +54,7 @@
                 <p v-if="item.description" class="apos-widget__help">
                   {{ $t(item.description) }}
                 </p>
-              </div>
+              </button>
```
Comment (ETLaurent) [inline modules/@apostrophecms/area/ui/apos/components/AposAreaExpandedMenu.vue:57]: make items focusable

File: modules/@apostrophecms/area/ui/apos/components/AposAreaExpandedMenu.vue [hunk 3]
```diff
@@ -239,6 +239,12 @@ export default {
 }
 
 .apos-widget {
+  @include type-base;
+  padding: 0;
+  border: none;
+  background: none;
+  text-align: inherit;
+
```
Comment (ETLaurent) [inline modules/@apostrophecms/area/ui/apos/components/AposAreaExpandedMenu.vue:247]: adapt css so that buttons look like div, as before

File: modules/@apostrophecms/command-menu/ui/apos/components/AposCommandMenuShortcut.vue [hunk 4]
```diff
@@ -62,6 +62,7 @@ export default {
       modal: {
         busy: false,
         active: false,
+        trapFocus: false,
```
Comment (ETLaurent) [inline modules/@apostrophecms/command-menu/ui/apos/components/AposCommandMenuShortcut.vue:65]: we explicitly don't want to trap the focus for the command menu shortcut modal

File: modules/@apostrophecms/doc-type/ui/apos/components/AposDocEditor.vue [hunk 5]
```diff
@@ -399,9 +401,11 @@ export default {
       this.docFields.data = newInstance;
       this.prepErrors();
       this.docReady = true;
+      this.modal.triggerFocusRefresh++;
     } else {
-      this.$nextTick(() => {
-        this.loadNewInstance();
+      this.$nextTick(async () => {
+        await this.loadNewInstance();
+        this.modal.triggerFocusRefresh++;
```
Comment (ETLaurent) [inline modules/@apostrophecms/doc-type/ui/apos/components/AposDocEditor.vue:408]: refresh focus elements in the modal, to take into account elements to focus from the new instance

File: modules/@apostrophecms/image/ui/apos/components/AposMediaUploader.vue [hunk 6]
```diff
@@ -30,6 +34,7 @@
       :accept="accept"
       multiple="true"
       :disabled="disabled"
+      tabindex="-1"
```
Comment (ETLaurent) [inline modules/@apostrophecms/image/ui/apos/components/AposMediaUploader.vue:37]: the file input is hidden, so we make the drag&drop element focusable and fake-trigger the input file click when pressing space or enter on the drag&drop element,

File: modules/@apostrophecms/modal/ui/apos/components/AposModal.vue [hunk 7]
```diff
@@ -12,6 +12,8 @@
       aria-modal="true"
       :aria-labelledby="id"
       ref="modalEl"
+      @keydown="cycleElementsToFocus"
+      @focus.capture="storeFocusedElement"
```
Comment (ETLaurent) [inline modules/@apostrophecms/modal/ui/apos/components/AposModal.vue:16]: listeners coming from the AposFocusMixin

File: modules/@apostrophecms/modal/ui/apos/components/AposModal.vue [hunk 8]
```diff
@@ -187,12 +197,20 @@ export default {
...
+    // that might have been created or removed
+    // after an update, like an XHR call to get the
+    // pieces list in the Doc Manager modal, for instance.
+    triggerFocusRefresh (newVal) {
+      console.log('triggerFocusRefresh: ', newVal);
+      if (this.shouldTrapFocus) {
+        this.$nextTick(this.trapFocus);
+      }
+    }
+  },
```
Comment (ETLaurent) [inline modules/@apostrophecms/modal/ui/apos/components/AposModal.vue:210]: this is the trick to refresh new elements to focus after some update inside a modal

File: modules/@apostrophecms/modal/ui/apos/components/AposModal.vue [hunk 9]
```diff
@@ -217,6 +235,7 @@ export default {
       // pop doesn't quite suffice because of race conditions when
       // closing one and opening another
       apos.modal.stack = apos.modal.stack.filter(modal => modal !== this);
+      this.focusLastModalFocusedElement();
```
Comment (ETLaurent) [inline modules/@apostrophecms/modal/ui/apos/components/AposModal.vue:238]: when we leave a modal, we want to focus back the stored focused element from the previous modal

File: modules/@apostrophecms/modal/ui/apos/components/AposModal.vue [hunk 10]
```diff
@@ -232,47 +251,32 @@ export default {
...
-        } else {
-          // If tab key is pressed
-          if (document.activeElement === lastFocusableElement) {
-            // Add focus for the first focusable element
-            firstFocusableElement.focus();
-            e.preventDefault();
-          }
-        }
+      function addExcludingAttributes(element) {
+        return `${element}:not([tabindex="-1"]):not([disabled]):not([type="hidden"]):not([aria-hidden])`;
```
Comment (ETLaurent) [inline modules/@apostrophecms/modal/ui/apos/components/AposModal.vue:276]: this is important, avoid focusing elements that are hidden/disabled or that have a negative tabindex

File: modules/@apostrophecms/modal/ui/apos/components/AposModal.vue [hunk 11]
```diff
@@ -232,47 +251,32 @@ export default {
...
-      const firstFocusableElement = focusables[0];
-      const lastFocusableElement = focusables[focusables.length - 1];
 
-      modalEl.addEventListener('keydown', cycleFocusables);
+      const selector = elementSelectors
+        .map(addExcludingAttributes)
+        .join(', ');
 
-      firstFocusableElement.focus();
+      this.elementsToFocus = [ ...this.$refs.modalEl.querySelectorAll(selector) ]
```
Comment (ETLaurent) [inline modules/@apostrophecms/modal/ui/apos/components/AposModal.vue:267]: update the list of elements to focus, so that the `cycleElementsToFocus` listener from the mixin has the first and the last elements updated, which is necessary to trap the focus

File: modules/@apostrophecms/modal/ui/apos/components/AposModalConfirm.vue [hunk 12]
```diff
@@ -95,8 +95,7 @@ export default {
         active: false,
         type: 'overlay',
         showModal: false,
-        disableHeader: true,
-        trapFocus: true
```
Comment (ETLaurent) [inline modules/@apostrophecms/modal/ui/apos/components/AposModalConfirm.vue:99]: remove this option since trapFocus is enabled by default

File: modules/@apostrophecms/page/ui/apos/components/AposPagesManager.vue [hunk 13]
```diff
@@ -221,6 +225,8 @@ export default {
     // Get the data. This will be more complex in actuality.
     this.modal.active = true;
     await this.getPages();
+    this.modal.triggerFocusRefresh++;
```
Comment (ETLaurent) [inline modules/@apostrophecms/page/ui/apos/components/AposPagesManager.vue:228]: take into account retrieved pages elements to add them in the list of elements to focus.

File: modules/@apostrophecms/ui/ui/apos/components/AposButtonSplit.vue [hunk 14]
```diff
@@ -133,11 +139,22 @@ export default {
...
-      setTimeout(() => {
-        this.$refs.choices[0].focus();
-      }, 200);
+    trapFocus() {
+      const selectedElementIndex = this.menu.findIndex(i => i.action === this.action) || 0;
+
+      // use map to keep items order:
+      this.elementsToFocus = this.menu.map(
+        i => this.$refs.choices.find(choice => choice.value === i.action)
+      );
```
Comment (ETLaurent) [inline modules/@apostrophecms/ui/ui/apos/components/AposButtonSplit.vue:148]: with `this.elementsToFocus = this.$refs.choices`, elements were not following the order in the DOM, leading to elements unfocusable and losing the focus.

File: modules/@apostrophecms/ui/ui/apos/scss/global/_admin.scss [hunk 15]
```diff
@@ -12,3 +12,12 @@
...
 }
+
+// TODO: remove these forced focus styles after keyboard accessibility is done.
+// This is useful for testing to visually see focused elements
+// that may not have the browser-native blue outline when focused.
+//
+// *:focus, *:focus-visible {
+//   outline: blue solid 2px!important;
+//   background-color: rgba(0,0,255,0.3)!important;
+// }
```
Comment (ETLaurent) [inline modules/@apostrophecms/ui/ui/apos/scss/global/_admin.scss:23]: let's keep this commented while the work on keyboard accessibility is not done.

File: modules/@apostrophecms/modal/ui/apos/components/AposModal.vue [hunk 16]
```diff
@@ -187,12 +197,19 @@ export default {
...
-      });
+    // Simple way to re-trigger focusable elements
+    // that might have been created or removed
+    // after an update, like an XHR call to get the
+    // pieces list in the AposDocsManager modal, for instance.
+    triggerFocusRefresh (newVal) {
+      if (this.shouldTrapFocus) {
+        this.$nextTick(this.trapFocus);
+      }
+    }
```
Comment (ETLaurent) [inline modules/@apostrophecms/modal/ui/apos/components/AposModal.vue:208]: this is the trick to refresh new elements to focus after some update that happened inside a parent component (AposDocEditor, AposDocsManager, AposPagesManager...)

File: CHANGELOG.md [hunk 17]
```diff
@@ -1,5 +1,13 @@
 # Changelog
 
+## UNRELEASED
+
+### Adds
+
+* Accessibility improved for navigation inside modals and various UI elements.  
+Pages/Docs Manager and Doc Editor modal now have a better keyboard accessibility by
+keeping the focus on elements inside modals and giving it back to their parent modal when closed.
+
```
Comment (ETLaurent) [inline CHANGELOG.md:10]: feel free to rephrase it to a better English 😶

Comment (boutell) [reply CHANGELOG.md:10]: "a better" => "better"

"... better keyboard accessibility. They keep the focus on elements inside modals and give it back..."

File: modules/@apostrophecms/area/ui/apos/components/AposAreaExpandedMenu.vue [hunk 18]
```diff
@@ -24,7 +24,7 @@
                 }`
               ]"
             >
-              <div
+              <button
```
Comment (boutell) [inline modules/@apostrophecms/area/ui/apos/components/AposAreaExpandedMenu.vue:27]: 💯 for semantic things (and their accessibility benefits)

File: modules/@apostrophecms/modal/ui/apos/components/AposModal.vue [hunk 19]
```diff
@@ -232,47 +250,26 @@ export default {
       this.$emit('esc');
     },
     trapFocus () {
-      // Adapted from https://uxdesign.cc/how-to-trap-focus-inside-modal-to-make-it-ada-compliant-6a50f9a70700
-      // All the elements inside modal which you want to make focusable.
-      const focusableElements = [
-        'button',
+      const elementSelectors = [
+        '[tabindex]',
```
Comment (boutell) [inline modules/@apostrophecms/modal/ui/apos/components/AposModal.vue:254]: Right, and we'll put this on tiptap presumably (I know tiptap will have its own constellation of accessibility problems too and that's OK for another ticket)

**Review by boutell — APPROVED** [review]
I think this is a good implementation. Thank you. There will be a need for more `triggerFocusRefresh` changes each time we add a new "toggle" component or similar, but I see no way around that.

**Review by boutell — APPROVED** [review]

**Review by stuartromanek — APPROVED** [review]
Wow, huge improvement @ETLaurent, nicely done.

Some thoughts I had while demoing, none of which are blockers

- As a follow-on in this spirit we should make it so our mechanism for opening context menus changes the focus to the element the context menu has revealed.
- [Media Library] Tabbing through the media library doesn’t cycle through the media itself, it probably should.
- [UX??] Not sure about this but when modals open we focus the first element, which is usually a cancel button. ‘Cancel’ is not the primary operation of the thing you opened (would New be? unsure..) and Cancel is always accessible by `esc` key.. Wondering if modals could have an optional param to focus a specific element within them on open and jumping to that on first open. /shrug

**Comment by ETLaurent** [comment]
> Wow, huge improvement @ETLaurent, nicely done.
> 
> Some thoughts I had while demoing, none of which are blockers
> 
> * As a follow-on in this spirit we should make it so our mechanism for opening context menus changes the focus to the element the context menu has revealed.
> * [Media Library] Tabbing through the media library doesn’t cycle through the media itself, it probably should.
> * [UX??] Not sure about this but when modals open we focus the first element, which is usually a cancel button. ‘Cancel’ is not the primary operation of the thing you opened (would New be? unsure..) and Cancel is always accessible by `esc` key.. Wondering if modals could have an optional param to focus a specific element within them on open and jumping to that on first open. /shrug

Thanks! All of your thoughts are right, and the first two are already in logged in tickets :)

You're making an interesting point about not focusing the first found element, but an element that we would set as "prioritized". Probably doable with tabindexes!
